## User-User Collaborative Filtering Assignment

### Part 1 . Without Normalization 
* data : 25 users x 100 movie matrix of rating 
* sheet1 ) rating matrix with movies as rows and users as columns
* sheet2 ) rating matrix with movies as colums and users as rows
* sheet3 ) correlation matrix (empty)

1. import dataset

In [1]:
import numpy as np
import pandas as pd

In [2]:
!pip install xlrd

In [3]:
import xlrd

In [5]:
df = pd.read_excel("UUCF Assignment Spreadsheet.xls",
                  sheet_name = 'movie-row')

In [6]:
df

Unnamed: 0  1648  5136  918  2824  \
0       11: Star Wars: Episode IV - A New Hope (1977)   NaN   4.5  5.0   4.5   
1                             12: Finding Nemo (2003)   NaN   5.0  5.0   NaN   
2                             13: Forrest Gump (1994)   NaN   5.0  4.5   5.0   
3                          14: American Beauty (1999)   NaN   4.0  NaN   NaN   
4   22: Pirates of the Caribbean: The Curse of the...   4.0   5.0  3.0   4.5   
..                                                ...   ...   ...  ...   ...   
95                       9806: The Incredibles (2004)   3.5   5.0  3.5   NaN   
96                 10020: Beauty and the Beast (1991)   3.0   5.0  NaN   NaN   
97                                36657: X-Men (2000)   NaN   4.5  NaN   NaN   
98                     36658: X2: X-Men United (2003)   3.5   4.0  NaN   NaN   
99                            36955: True Lies (1994)   NaN   NaN  NaN   NaN   

    3867  860  3712  2968  3525  ...  3556  5261  2492  5062  2486  4942  \
0    4.0  4.0   NaN   5.0   4.0  ...   4.0   NaN   4.5   4.0   3.5   NaN   
1    4.0  4.0   4.5   4.5   4.0  ...   4.0   NaN   3.5   4.0   2.0   3.5   
2    4.5  4.5   NaN   5.0   4.5  ...   4.0   5.0   3.5   4.5   4.5   4.0   
3    NaN  NaN   4.5   2.0   3.5  ...   4.0   NaN   3.5   4.5   3.5   4.0   
4    4.0  2.5   NaN   5.0   3.0  ...   3.0   1.5   4.0   4.0   2.5   3.5   
..   ...  ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...   
95   3.0  3.5   4.5   5.0   3.5  ...   3.5   0.5   5.0   3.5   3.0   NaN   
96   4.0  3.5   NaN   2.0   4.0  ...   4.0   NaN   4.0   2.0   2.5   NaN   
97   4.0  NaN   4.5   5.0   4.0  ...   3.5   NaN   3.0   4.0   3.5   NaN   
98   3.5  NaN   4.5   4.5   4.0  ...   3.5   NaN   2.0   NaN   3.0   4.0   
99   3.0  3.5   4.0   1.5   3.0  ...   NaN   NaN   4.0   NaN   1.5   4.0   

    2267  4809  3853  2288  
0    NaN   NaN   NaN   NaN  
1    NaN   NaN   NaN   3.5  
2    3.5   4.5   3.5   3.5  
3    NaN   3.5   NaN   NaN  
4    NaN   5.0   NaN   3.5  
..   ...   ...   ...   ...  
95   NaN   4.5   NaN   3.5  
96   NaN   4.0   NaN   NaN  
97   NaN   NaN   3.0   4.0  
98   NaN   3.5   NaN   4.0  
99   NaN   NaN   NaN   NaN  

[100 rows x 26 columns]

2. Complete the user-by-user correlations matrix. To check your math, note that the correlation between users 1648 and 5136 is 0.40298, and the correlation between users 918 and 2824 is -0.31706. All correlations should be between -1 and 1, and the diagonal should be all 1's (since they are self-correlations).

3. dentify the top 5 neighbors (the users with the 5 largest, positive correlations) for users 3867 and 89. For example, if the target user were #3712, the closest neighbors are 2824 (corr: 0.46291), 3867 (corr: 0.400275), 5062 (corr: 0.247693), 442 (corr: 0.22713), and 3853 (corr: 0.19366). Don't forget to exclude the target user (corr: 1.0) from your possible selections.

4. Create a new worksheet in your spreadsheet, and use it to compute the predictions for each movie for users 3867 and 89 by taking the correlation-weighted average of the ratings of the top-five neighbors (for each target user) for each movie. The formal formula for correlation-weighted average is $\sum_{n=1}^{5} r_n w_n \over \sum_{n=1}^{5} w_n$
 . Remember, you will need to make sure that your weight for each contributed rating is the user-user correlation when that neighbor has rated the movie, but 0 when the neighbor has not rated the movie).

5. Submit 12 values for this assignment as indicated below. You will be submitting the top three movie IDs and the predicted ratings (to three decimal places) for each user. In a real recommender system, we’d be excluding movies the user has already rated, but do not do this here. Indeed, you should look to see what the user’s rating (if any) is for the top-recommended movies. For example, if the user ID was 3712, the correct submission would be:
* Top Movie: 641 Prediction: 5.000

* 2nd Movie: 603 Prediction: 4.856

* 3rd Movie: 105 Prediction: 4.739

* And if the user ID were 3525, there would be a three-way tie among: Movies: 238, 194, and 38 (all with a prediction of 5.000)